### ISSUE TITLE
Add constructor `simd<T, Abi>{T...}`

### ISSUE THREAD

**Issue by bernhardmgruber** [mattkretz/std-simd-feedback#24]
While it is not immediately obvious how many elements a `simd` will have, in some cases, like with the `fixed_size` abi, it is. So it would be great if we could have a constructor that allows to initialize a `simd` of size `N` with exactly `N` constructor arguments:

```c++
stdx::fixed_size_simd<double, 4> simd{0.0, 11.0, 22.0, 33.0};
```

I leave it up to you whether to also add such a constructor for `simd`s of other abis.

**Comment by mattkretz** [comment]
I'm torn on this issue. FWIW your specific example should be like
```c++
constexpr auto x = any_simd_type::iota * 11;
```
Sadly, `iota` wasn't accepted by WG21 for the TS, so you have to write `any_simd_type([](auto i) -> T { return i; }` instead of `iota`. Feel free to give feedback that you want `iota` for the IS merge :wink:.

That said, I admit are still cases where an initializer list can be useful, even if "dangerous" for portability. Questions:

* Should `fixed_size_simd<int, 4> x{1, 2, 3}` be valid? `{1, 2, 3, 0}` or `{1, 2, 3, 1}`?
* Should `fixed_size_simd<int, 4> x{1, 2, 3, 4, 5}` be valid? `{1, 2, 3, 4}`?
* Should `native_simd<int> x{1, 2}` be valid? `{1, 2, 0, 0, 0, 0, 0, 0}` or `{1, 2, 1, 2, 1, 2, 1, 2}` (for size 8)?

**Comment by bernhardmgruber** [comment]
> FWIW your specific example should be like
> 
> ```c++
> constexpr auto x = any_simd_type::iota * 11;
> ```

OK, that seems really useful! Fortunately, we could still propose `any_simd::iota` at a later time.

> * Should `fixed_size_simd<int, 4> x{1, 2, 3}` be valid? `{1, 2, 3, 0}` or `{1, 2, 3, 1}`?
> * Should `fixed_size_simd<int, 4> x{1, 2, 3, 4, 5}` be valid? `{1, 2, 3, 4}`?
> * Should `native_simd<int> x{1, 2}` be valid? `{1, 2, 0, 0, 0, 0, 0, 0}` or `{1, 2, 1, 2, 1, 2, 1, 2}` (for size 8)?

I would be either conservative and disallow the constructor argument count to be different than the SIMD size, or allow less arguments with the unspecified arguments to be value initialized in the SIMD. The latter is how arrays and aggregates behave, so there is precedence.

> even if "dangerous" for portability

That was also my concern. If someone writes `native_simd<int> x{1, 2, 3, 4}` in their program and it compiles, they could falsely assume their code is correct and portable. But the constructor would fail to compile on an architecture with a shorter native SIMD length or be filled with trailing zeros on a longer SIMD length.

In any case, the constructors can be added later, after the base functionality is merged. And maybe it's safe to propose only the one with an exact match of arguments and then maybe even enable it for the `fixed_size` ABI only.

**Comment by mattkretz** [comment]
> In any case, the constructors can be added later, after the base functionality is merged.

Yes, however it should probably be done in the C++26 timeframe. Otherwise you need to be certain that is cannot break existing code. And C++ has ways to surprise, especially with initializer lists.

> maybe it's safe to propose only the one with an exact match of arguments and then maybe even enable it for the fixed_size ABI only.

I think I agree to go with fixed_size only and consequently asking for an exact match. But it needs to be explored, discussed wrt. alternatives, explained why that's the best, etc. I'll be happy to nitpick your paper to make it stronger, if you find time to work on it. (please do, I think it's a net win)

**Comment by mattkretz** [comment]
Oh, it just occurred to me that there's a relation to what I'd like to propose wrt. `span`s of static extent:

```c++
void f(std::span<float, 4> x) {
  simd v = x; // CTAD -> fixed_size_simd<float, 4>
}
```

Example: https://godbolt.org/z/K7919qecP

**Comment by bernhardmgruber** [comment]
> > In any case, the constructors can be added later, after the base functionality is merged.
> 
> Yes, however it should probably be done in the C++26 timeframe. Otherwise you need to be certain that is cannot break existing code. And C++ has ways to surprise, especially with initializer lists.

So I may interpret this as: `std::simd` will make it in C++26? :) :D

> > maybe it's safe to propose only the one with an exact match of arguments and then maybe even enable it for the fixed_size ABI only.
> 
> I think I agree to go with fixed_size only and consequently asking for an exact match. But it needs to be explored, discussed wrt. alternatives, explained why that's the best, etc. I'll be happy to nitpick your paper to make it stronger, if you find time to work on it. (please do, I think it's a net win)

I put it on my TODO list! As a PhD student in their third year, I have to focus a few other things now before I can get back to that.

**Comment by bernhardmgruber** [comment]
> And C++ has ways to surprise, especially with initializer lists.

Just FYI, was thinking about something like:
```c++
template <typename T>
struct simd {
...
  template <typename... Us>
  constexpr simd(Us... us) requires (sizeof...(Us) == size) // and same requirements as the single value ctor for each Us
      : data{us...} {}
};
```
Godbolt: https://godbolt.org/z/n5c6eo1vP

So, no `std::initializer_list<T>` is involved.

**Comment by mattkretz** [comment]
Yes, also `sizeof...(Us) > 1` to avoid the broadcast ctor in the scalar case. And conversion requirements. Then I think this should work without breaking anything.

> I put it on my TODO list

the black hole without a bad conscience :stuck_out_tongue_winking_eye:. No, I fully understand.

> So I may interpret this as: std::simd will make it in C++26?

Almost. Interpret as: "several people will try their best to put std::simd with improvements and extensions into C++26".

**Comment by bernhardmgruber** [comment]
> > So I may interpret this as: std::simd will make it in C++26?
> 
> Almost. Interpret as: "several people will try their best to put std::simd with improvements and extensions into C++26".

I am happy to hear you don't need to push this through alone! If there is anything smallish that I can help with, please tell me! Thank you for the great work!

**Comment by danieltowner** [comment]
> > And C++ has ways to surprise, especially with initializer lists.
> 
> Just FYI, was thinking about something like:
> 
> ```c++
> template <typename T>
> struct simd {
> ...
>   template <typename... Us>
>   constexpr simd(Us... us) requires (sizeof...(Us) == size) // and same requirements as the single value ctor for each Us
>       : data{us...} {}
> };
> ```
> 
> Godbolt: https://godbolt.org/z/n5c6eo1vP
> 
> So, no `std::initializer_list<T>` is involved.

That's a neat way to do it. In the Intel implementation of simd we have an initialiser_list constructor but it doesn't handle constexpr properly, which is a pain because if you are initialising a simd with some constants it probably wants to be constexpr. Mostly the compiler generates what it would have if it really had been constexpr, but the simd can't be formally used in a constexpr expression that follows on.

**Comment by danieltowner** [comment]
> > In any case, the constructors can be added later, after the base functionality is merged.
> 
> Yes, however it should probably be done in the C++26 timeframe. Otherwise you need to be certain that is cannot break existing code. And C++ has ways to surprise, especially with initializer lists.
> 
> > maybe it's safe to propose only the one with an exact match of arguments and then maybe even enable it for the fixed_size ABI only.
> 
> I think I agree to go with fixed_size only and consequently asking for an exact match. But it needs to be explored, discussed wrt. alternatives, explained why that's the best, etc. I'll be happy to nitpick your paper to make it stronger, if you find time to work on it. (please do, I think it's a net win)

In the Intel implementation which has a single-element broadcast constructor and also an initialiser list it gives the same behaviour as an array for consistency, with zero filling:

std::array<int, 5> a = {1, 2}; // Creates [1, 2, 0, 0, 0]
simd<int, 5> s_a = {1, 2}; // Also creates [1, 2, 0, 0, 0]

And if there is only a single element we get a match between the two as well:

std::array<int, 5> a = {99}; // Creates [99, 0, 0, 0, 0]
simd<int, 5> s = {99}; // Also creates [99, 0, 0, 0, 0]

But if we remove the initialiser list and switch to a multi-element constructor with lax counting and zero-fill instead:

std::array<int, 5> a = {99}; // Creates [99, 0, 0, 0, 0]
simd<int, 5> s = {99}; // Creates [99, 99, 99, 99, 99] because broadcast ctor is a better match - surprising outcome

So it seems like we either:
- have an initializer list which can differentiate between the use cases, give consistency against other containers, but can't be constexpr
- have a multi-element constructor which enforces sizing to ensure that we don't trip over an initialisation from a list of 1, and which gives us constexpr properties.

I think possibly the latter is better, simply because I rate constexpr behaviour as being more important (i.e., the main time you would be initialising a simd with a few known constants is when it really deserves to be a constexpr). I think I'd allow it on any simd abi though - even a native_simd - because it might get used inside an if constexpr which happens to specialise for a known native platform. 

https://godbolt.org/z/qcjWGEjnP

**Comment by mattkretz** [comment]
`constexpr` `initializer_list` simd constructor with zero-fill:
```c++
  constexpr simd(const std::initializer_list<T> &list)
      : data([&list]<std::size_t... Is>(std::index_sequence<Is...>) {
          return V{(Is < list.size() ? list.begin()[Is] : 0)...};
        }(std::make_index_sequence<N>())) {}
```

I'm not happy about:
```c++
simd<int> a(1);    // [1, 1, 1, 1]
simd<int> b{1};    // [1, 0, 0, 0]
simd<int> c = {1}; // [1, 0, 0, 0]
```

It's too similar to produce different results. The user who wants to avoid the vexing parse gets different behavior. Not good.

**Comment by mattkretz** [comment]
It would be possible to get the following behavior (with a `constexpr` `initializer_list` ctor):
```c++
simd<int> a(1);    // [1, 1, 1, 1]
simd<int> b{1};    // [1, 1, 1, 1]
simd<int> c = {1}; // [1, 1, 1, 1]
simd<int> d{1, 0};    // [1, 0, 0, 0]
simd<int> e = {1, 0}; // [1, 0, 0, 0]
```

What do you think?

**Comment by bernhardmgruber** [comment]
> It would be possible to get the following behavior (with a constexpr initializer_list ctor) ...

I like the consistency. It's good and would serve my needs. But what is the advantage of a `constexpr` ctor with `initializer_list` over a variadic ctor? I think the variadic ctor could handle:

```c++
int n = ...;
simd<int> a{n, n, n, n};
```
And the `initializer_list` probably not, because `n` is not a constant expression. (I have not tried and verified this)

**Comment by mattkretz** [comment]
> I like the consistency. It's good and would serve my needs. But what is the advantage of a `constexpr` ctor with `initializer_list` over a variadic ctor?

The `constexpr` is not a restriction to `consteval`, so `a{n, n, n, n}` still works (and is actually translated to a broadcast).

It would just be sad if constant expressions of type `simd` could not be formed from initializer lists.

## which ctor is better?

The `simd(Us...)` ctor is more complex because of the individual type deductions per argument. You need to constrain (and convert) them all in the `simd` implementation. This ctor will have to be considered on overload resolution every single time any `simd` ctor is needed. Whereas an `initializer_list` ctor requires uniform types (I'd go with the element type, no conversions), so the issue must be solved by the caller. Which one is better? Pros and cons for both, I guess. Explore... ;)

(modified example: https://godbolt.org/z/8K3EeYdbz)

**Comment by danieltowner** [comment]
A common case I keep coming across is wanting to build a simd with a few known constants. I've added a deduction guide to make this easier. It allows something like:

simd values = {5.f, 6.f, 7.f};

which in this case deduces that fixed_size_simd<float, 3> is the most appropriate size. With the deduction guide fixing the size of the simd to exactly the length of the list it then allows us to define a constructor which takes exactly the right number of arguments, and avoid the question of what to do with unspecified elements.

**Comment by danieltowner** [comment]
> Sadly, `iota` wasn't accepted by WG21 for the TS, so you have to write `any_simd_type([](auto i) -> T { return i; }` instead of `iota`. Feel free to give feedback that you want `iota` for the IS merge 😉.

I've now mentioned iota at Kona '22 and Issaquah '23 with no objections raised on either occasion. Can we just proceed with including it?

**Comment by mattkretz** [comment]
:+1: for including `iota`. I'd be happy for a PR to P1928 (even if the paper doesn't compile) that mentions the addition in Section 4 and adds it to the wording.

Do we all agree on `simd<T>::iota`? Or should it be `std::iota<simd<T>>`? The latter would allow `iota<float>` (equal to 0) and enable more generic code. `std::iota<std::array<int, 1024>>`...? There's still a design space where we need to nail it down somewhere.

**Comment by danieltowner** [comment]
I do like the  more generic version.

**Comment by bernhardmgruber** [comment]
> Do we all agree on `simd<T>::iota`? Or should it be `std::iota<simd<T>>`?

I assume this collides with the [`std::iota` algorithm](https://en.cppreference.com/w/cpp/algorithm/iota). Furthermore, I believe grabbing a name for this facility in `std::` will be harder than using `iota` as a member. But I have no preference, whatever you deem best.

**Comment by mattkretz** [comment]
Yes, it conflicts with `std::iota` defined as a function. Unless we make it `std::iota<T>()`, i.e. a function overload. It doesn't collide in the "two different concepts with the same name" sense. It collides in the "two different ways to use something with the same name" sense.

Choices:
1. `std::simd<T>::iota`
2. `std::iota<T>()`
3. `std::iota_v<T>`

The choices (2) and (3) can be generalized to more types: arithmetic/numeric types, simd<arithmetic>, std::array, C-array, more?

At this point I have a slight preference for (3) over (2). I'd take (1) only if WG21 objects to the others.

Variable templates without `_v` suffix are still weird sometimes. Like, "I need to put parens here".

**Comment by mattkretz** [comment]
Like this: https://github.com/mattkretz/simd-prototyping/blob/main/iota.h

```c++
auto a = std::iota_v<std::simd<float>>;
auto b = std::iota_v<float>;
auto c = std::iota_v<std::array<float, 4>>;
auto d = std::iota_v<float[4]>;
```
```asm
d:
	.quad	std::__detail::__iota_array<float, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul, 3ul> >::__data  # 0x0000000000000000
---------------------------------------------
std::__detail::__iota_array<float, std::integer_sequence<unsigned long, 0ul, 1ul, 2ul, 3ul> >::__data:
	.long	          0  # 0x00000000
	.long	 1065353216  # 0x3f800000
	.long	 1073741824  # 0x40000000
	.long	 1077936128  # 0x40400000
	.data
---------------------------------------------
c:
	.long	          0  # 0x00000000
	.long	 1065353216  # 0x3f800000
	.long	 1073741824  # 0x40000000
	.long	 1077936128  # 0x40400000
	.bss
---------------------------------------------
b:
	.zero	4 	#
	.data
---------------------------------------------
a:
	.long	          0  # 0x00000000
	.long	 1065353216  # 0x3f800000
	.long	 1073741824  # 0x40000000
	.long	 1077936128  # 0x40400000
	.long	 1082130432  # 0x40800000
	.long	 1084227584  # 0x40a00000
	.long	 1086324736  # 0x40c00000
	.long	 1088421888  # 0x40e00000
	.long	 1090519040  # 0x41000000
	.long	 1091567616  # 0x41100000
	.long	 1092616192  # 0x41200000
	.long	 1093664768  # 0x41300000
	.long	 1094713344  # 0x41400000
	.long	 1095761920  # 0x41500000
	.long	 1096810496  # 0x41600000
	.long	 1097859072  # 0x41700000
```

**Comment by danieltowner** [comment]
> At this point I have a slight preference for (3) over (2). I'd take (1) only if WG21 objects to the others.

Agreed. 3 is the neatest and most generically useful.

**Comment by mattkretz** [comment]
If we don't add this constructor then user will come up with:
```c++
simd<int> iota(std::array{0, 1, 2, 3, 4, 5, 6, 7}.begin());
```
That's so much worse. I know about this hack because I just needed it. :scream:

**Comment by danieltowner** [comment]
How about borrowing from the example here: https://en.cppreference.com/w/cpp/ranges/iota_view

```constexpr auto vec = std::views::iota(1, 5) | std::ranges::to<std::simd>();```

It's a little verbose but it is obvious in its intent.
